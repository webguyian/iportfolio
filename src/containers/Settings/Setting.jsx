import React, { Fragment } from 'react';
import { func, object } from 'prop-types';
import cx from 'classnames';
import { Route, Switch } from 'react-router-dom';

import Button from 'components/Button/Button';
import Link from 'components/Link/Link';
import Text from 'components/Text/Text';

import { useSetting } from 'modules/settings/hooks';

const Setting = props => {
  const { setting, onClick } = props;
  const [backBtnRef, actions, state] = useSetting(setting);
  const baseClass = 'setting';
  const activeClass = state.active ? `${baseClass}--active` : false;
  const backLink = state.active ? `/settings/${setting.id}` : '/settings';
  const backFn = state.active ? () => actions.onNavigate() : onClick;

  return (
    <div className={cx(baseClass, activeClass)}>
      <nav className={`${baseClass}-nav`}>
        <Link back to={backLink} onClick={backFn} ref={backBtnRef} />
      </nav>
      <div className={`${baseClass}-view-container`}>
        <Switch>
          {setting.credit && (
            <Route path={`/settings/${setting.id}/credit`} exact>
              <div className={`${baseClass}-subview`}>
                <pre className={`${baseClass}-credit`}>
                  {setting.credit.description}
                </pre>
              </div>
            </Route>
          )}
          <Route>
            <div className={`${baseClass}-view`}>
              <Text className={`${baseClass}-label`} element="h2">
                {setting.label} Settings
              </Text>
              {!setting.noStorage && (
                <Button
                  className={`${baseClass}-btn`}
                  modifier="anchor-block"
                  onClick={actions.onClearCache}
                  disabled={state.disabled}
                >
                  Clear Storage Cache
                </Button>
              )}
              {setting.credit && (
                <Fragment>
                  <Text element="h3" type="display">
                    Acknowledgements
                  </Text>
                  <Link
                    className={`${baseClass}-item ${baseClass}-item-link`}
                    to={`/settings/${setting.id}/credit`}
                    onClick={actions.onNavigate}
                  >
                    {setting.credit.label}
                  </Link>
                </Fragment>
              )}
              {setting.permissions && (
                <Fragment>
                  <Text element="h3" type="display">
                    Permissions
                  </Text>
                  <Text className={`${baseClass}-item`}>
                    {setting.permissions}
                  </Text>
                </Fragment>
              )}
            </div>
          </Route>
        </Switch>
      </div>
    </div>
  );
};

Setting.propTypes = {
  onClick: func.isRequired,
  setting: object.isRequired
};

export default Setting;
