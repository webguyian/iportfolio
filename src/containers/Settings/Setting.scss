.setting {
  @include stretch-column-nowrap;

  &-nav {
    padding: $gutter-s 0;
  }

  &-view-container {
    display: flex;
    width: 100%;
    height: 100%;
  }

  &-view {
    flex: 1 0 100%;
    width: 375px;
  }

  &-subview {
    flex: 0 1 100%;
    width: 375px;

    .setting--active & {
      flex: 1 0 100%;
    }
  }

  .ui-link--back {
    color: $blue-light;
  }

  .ui-text--display {
    margin-top: $gutter;
    margin-bottom: $gutter-xs;
    font-size: $font-size-m;
  }

  &-label {
    color: $gray;
    font-size: $font-size-m;
    font-variant: small-caps;
    text-transform: lowercase;
  }

  &-btn {
    @include settings-item;
    margin: $gutter-xs 0 $gutter;
    color: $red;

    &[disabled] {
      color: $gray-medium;
    }
  }

  &-item {
    @include settings-item;
    display: block;

    &-link {
      @include type-icon("\f054");
      position: relative;

      &::after {
        position: absolute;
        right: $gutter;
        color: $gray;
      }
    }
  }

  &-credit {
    width: 100%;
    max-height: 90%;
    padding: 0 $gutter-s;
    white-space: pre-line;
    overflow: auto;
  }
}
