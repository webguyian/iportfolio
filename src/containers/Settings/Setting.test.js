import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Setting from './Setting';

describe('<Setting />', () => {
  const props = {
    onClick: jest.fn(),
    setting: {
      icon: 'cameraIcon',
      id: 'camera',
      label: 'Camera',
      noStorage: true,
      permissions: 'Media Devices'
    }
  };

  it('renders correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Setting {...props} />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
