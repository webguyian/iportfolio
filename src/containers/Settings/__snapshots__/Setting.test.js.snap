// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Setting /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="setting"
  >
    <nav
      class="setting-nav"
    >
      <mock-link
        back="true"
        to="/settings"
      />
    </nav>
    <div
      class="setting-view-container"
    >
      <div
        class="setting-view"
      >
        <h2
          class="ui-text ui-text--body setting-label"
        >
          Camera Settings
        </h2>
        <h3
          class="ui-text ui-text--display"
        >
          Permissions
        </h3>
        <span
          class="ui-text ui-text--body setting-item"
        >
          Media Devices
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
